packages/controls/src/controls/group/group.ts

```diff
@@ -0,0 +1,310 @@
+    if (data.type !== Definition.type)
+      throw new Error(
+        `Group: expected '${Definition.type}', got '${data.type}'`,
+      )
+
+    const {
+      config: { props, ...config },
+    } = z
+      .object({
+        type: z.literal(Definition.type),
+        config: z.object({
+          label: z.string().optional(),
+          preferredLayout: z
+            .union([
+              z.literal(Definition.Layout.Inline),
+              z.literal(Definition.Layout.Popover),
+            ])
+            .optional(),
+          props: z.record(z.string(), SerializationSchema.deserializedRecord),
+        }),
+      })
+      .parse(data)
+
+    const deserializedProps = mapValues(props, (itemDef) => {
+      return deserializeCallback(itemDef)
+    })
+
+    return Group({
+      props: deserializedProps,
+      ...config,
+    })
+  }
+
+  constructor(
+    config: C,
+    readonly dataType:
+      | typeof Definition.v1DataType
+      | typeof ShapeV2Definition.v1DataType,
+  ) {
+    super(config)
+  }
+
+  private get dataSignature() {
+    return { [ControlDataTypeKey]: this.dataType }
+  }
+
+  get propDefs() {
+    return this.config.props
+  }
+
+  get controlType() {
+    return Definition.type
+  }
+
+  get schema() {
+    const version = z.literal(1)
+    const type = z.literal(Definition.type)
+
+    const keys = mapValues(this.propDefs, (def) => def.schema)
+
+    const propsData = z.object(
+      mapValues(keys, (key) => key.data),
+    ) as SchemaType<PropsData<C>>
+
+    const versionedData = z.object({
+      [ControlDataTypeKey]: z.union([
+        z.literal(Definition.v1DataType),
+        z.literal(ShapeV2Definition.v1DataType),
+      ]),
+      value: propsData,
+    }) as SchemaType<VersionedData<C>>
+
+    const data = z.union([propsData, versionedData])
+
+    const value = z.object(mapValues(keys, (key) => key.value)) as SchemaType<
+      ValueType<C>
+    >
+
+    const resolvedValue = z.object(
+      mapValues(keys, (key) => key.resolvedValue),
+    ) as SchemaType<ResolvedValueType<C>>
+
+    const config = z.object({
+      label: z.string().optional(),
+      preferredLayout: z.union([
+        z.literal(Definition.Layout.Inline),
+        z.literal(Definition.Layout.Popover),
+      ]),
+      props: z.object(mapValues(keys, (key) => key.definition)),
+    })
+
+    const definition = z.object({
+      type,
+      config,
+    })
+
+    return { type, data, value, resolvedValue, definition, version }
+  }
+
+  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
+    return safeParse(this.schema.data, data)
+  }
+
+  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
+    if (data == null) return undefined
+
+    const propsData = Definition.propsData(data)
+
+    return mapValues(this.propDefs, (def, key) => def.fromData(propsData[key]))
+  }
+
+  propsToData(propsData: PropsData<C>): DataType<C> {
+    return {
+      ...this.dataSignature,
```

Comment:
`propsToData` always returns a versioned data tagged with the current data signature, either `Group`'s or `ShapeV2`'s.

Reply:
This is the only place where we "convert" the data.